### PR TITLE
Add new unit and integration tests

### PR DESCRIPTION
**Pull request by sachanub**
*Issue #, if available:*

*Description of changes:*

The objective of this PR is to add new unit and integration tests. Specifically, we include the following:

- Unit test for `ts_environment.py`: **`test/unit/test_ts_environment.py`**
    - Added unit test to create a mock TorchServe environment and check the values of the environment variables.
- SageMaker integration tests for MNIST inductor: **`test/integration/sagemaker/test_mnist_inductor.py`**
    - **`test/resources/mnist/model_inductor`**: 
        - Included model files from DLC repo i.e. `code/mnist.py` and `torch_model.pth`
    - [**`test/integration/__init__.py`**](https://github.com/aws/sagemaker-pytorch-inference-toolkit/blob/master/test/integration/__init__.py): 
        - Added logic to define variables for MNIST inductor script and MNIST inductor model artifact.
    - **`test/integration/sagemaker/test_mnist_inductor.py`**:
        - Added CPU test to create and test endpoint with ml.c5.9xlarge instance type.
        - Added GPU test to create and test endpoints with ml.p3.2xlarge, ml.g4dn.4xlarge and ml.g5.4xlarge instance types. 
- SageMaker integration tests for multi-model endpoint (MME): **`test/integration/sagemaker/test_multi_model_endpoint_sagemaker.py`**
    - **`test/resources/mme`**: 
        - Copied ResNet-18 model artifact from `test/resources/resnet18/default_model`.
        - Copied traced ResNet-18 model artifact from `test/resources/resnet18/default_traced_resnet`.
    - [**`test/integration/__init__.py`**](https://github.com/aws/sagemaker-pytorch-inference-toolkit/blob/master/test/integration/__init__.py):
        - Add logic to create model artifacts for the MME tests.
    - **`test/integration/sagemaker/test_multi_model_endpoint_sagemaker.py`**:
        - Added test to check only ResNet-18 model is available on the multi-model endpoint.
        - Added test to check only traced ResNet-18 model is available on the multi-model endpoint.
        - Added test to check both models are available on the multi-model endpoint.
        - Added test to no models are available on the multi-model endpoint.
        - Added test to invoke both the ResNet-18 and traced ResNet-18 models and verify output length.
- Checks for verifying that dependencies from `requirements.txt` are installed:
    - **`requirements.txt`** file: This file has the `transformers` dependency.
        - `test/resources/mnist/model_cpu/code/requirements.txt`.
    - Added logic to try and import `transformers` in the following script:
        - `test/resources/mnist/model_cpu/code/mnist.py`.
    - [**`test/integration/__init__.py`**](https://github.com/aws/sagemaker-pytorch-inference-toolkit/blob/master/test/integration/__init__.py):
        - Added logic to include requirements.txt in MNIST tar file.
    - [**`test/utils/file_utils.py`**](https://github.com/aws/sagemaker-pytorch-inference-toolkit/blob/master/test/utils/file_utils.py):
        - Added logic to include `requirements.txt` in tar file creation. 
----------------------------------------------------------------------------------------------------------------------------

**UPDATE**:

- Refactored [**`test/integration/__init__.py`**](https://github.com/aws/sagemaker-pytorch-inference-toolkit/blob/master/test/integration/__init__.py) to obtain model information from `test/integration/all_models_info.json` and iteratively create attributes for the model directory, model script name and model tar file.

Added the following local integration tests:

- Local integration test for multi-model endpoint:
    - **`test/integration/sagemaker/test_multi_model_endpoint_local.py`**:
        - Added test to send a successful ping.
        - Added test to check that list of models is empty.
        - Added test to check that the ResNet-18 and traced ResNet-18 models are successfully loaded.
        - Added test to check that the ResNet-18 model is successfully unloaded.
        - Added test to assert failure on trying to unload a model which has already been unloaded i.e. ResNet-18.
        - Added test to assert failure on trying to load a model which has already been loaded i.e. traced ResNet-18.
        - Added test to invoke both the ResNet-18 and traced ResNet-18 models and verify output length.
- Local integration test to check that all GPU IDs are returned on multi-GPU host:
    - [**`test/integration/__init__.py`**](https://github.com/aws/sagemaker-pytorch-inference-toolkit/blob/master/test/integration/__init__.py):
        - Define directory path for `model_gpu_context`.
    - **`test/resources/model_gpu_context/code/inference.py`**:
        - Create custom `model_fn`, `input_fn`, `predict_fn` and `output_fn` to dynamically get all the GPU IDs from the `context` object and their corresponding PIDs and thread IDs and write them in different CSV files. Return dummy values from the functions.
    - **`test/integration/local/test_model_fn_context.py`**:
        - Added test to check all GPU IDs are returned by the `context` object in the `model_fn`, `input_fn`, `predict_fn` and `output_fn` functions by reading from the CSV files created in `test/resources/model_gpu_context/code/inference.py`.
        - Added test to check that for a given device ID, PID and thread ID are same for the `model_fn`, `input_fn`, `predict_fn`, `output_fn` functions.
   

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/unit/test_ts_environment.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Copyright 2019-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (chen3933) [inline test/unit/test_ts_environment.py:1]: You probably want to cover `is_env_set` as well : https://github.com/aws/sagemaker-pytorch-inference-toolkit/blob/master/src/sagemaker_pytorch_serving_container/ts_environment.py#L58-L64

Comment (sachanub) [reply test/unit/test_ts_environment.py:1]: Added test for `is_env_set()` in `test_ts_environment.py`.

**Comment by chen3933** [comment]
I notice we don't have local integration test. Is there any use case we want to cover in local integration test?

**Comment by chen3933** [comment]
I didn't see user script overwrites `input_fn` or `output_fn`. Do we want to add those tests?

**Comment by sachanub** [comment]
> I notice we don't have local integration test. Is there any use case we want to cover in local integration test?

Added the following local integration tests:
- MME tests.
- Test to check all GPU device IDs are dynamically returned by the `context` object in `model_fn`.

**Comment by sachanub** [comment]
> I didn't see user script overwrites `input_fn` or `output_fn`. Do we want to add those tests?

We define custom `input_fn` and `output_fn` for this test: https://github.com/aws/sagemaker-pytorch-inference-toolkit/blob/master/test/integration/local/test_serving.py#L87. This script defines custom `input_fn` and `output_fn` functions: https://github.com/aws/sagemaker-pytorch-inference-toolkit/blob/master/test/resources/mnist/model_cpu/code/call_model_fn_once.py

File: test/integration/local/test_model_context.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    output_fn_device_info = read_csv("output_fn_device_info.csv")
+    predict_fn_device_info = read_csv("predict_fn_device_info.csv")
+
+    assert len(input_fn_device_info) == len(output_fn_device_info) == len(predict_fn_device_info)
+
+    for input_fn_row, output_fn_row, predict_fn_row in zip(
+        input_fn_device_info, output_fn_device_info, predict_fn_device_info
+    ):
+
+        device_id_input_fn, pid_input_fn, threadid_input_fn = input_fn_row
```
Comment (chen3933) [inline test/integration/local/test_model_context.py:99]: We cannot guarantee the sequence of `input_fn_device_info.csv`, `output_fn_device_info.csv` and `predict_fn_device_info.csv` are the same, right?

Eg:
input_fn_device_info.csv
```
pid=0 device_id=0
pid=1 device_id=1
```

output_fn_device_info.csv
```
pid=1 device_id=1
pid=0 device_id=0
```

in above, the behavior is correct but test will fail.

Comment (chen3933) [reply test/integration/local/test_model_context.py:99]: Also, we want to check device_id consistency across all four functions, namely `model_fn`, `input_fn`, `output_fn` and `predict_fn`

File: test/resources/model_gpu_context/code/inference.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import torch
+
+def model_fn(model_dir, context):
+    script_dir = os.path.dirname(os.path.abspath(__file__))
+    file_path = os.path.join(script_dir, "device_info.txt")
+    
+    device = torch.device("cuda:" + str(context.system_properties.get("gpu_id")))
+    device_str = str(device)[-1]
+    with open(file_path, "a") as file:
+        file.write(device_str + "\n")
```
[the remark this answers is not in the document]
Comment (sachanub) [reply test/resources/model_gpu_context/code/inference.py]: Acknowledged. Fixed in latest commit.

File: test/integration/local/test_model_context.py (hunk 2 above)
Comment (sachanub) [reply test/integration/local/test_model_context.py]: Acknowledged. Fixed in latest commit.

**Review by chen3933 — APPROVED** [review]

File: test/resources/mnist/model_cpu/code/mnist.py [hunk 4]
```diff
@@ -16,6 +16,11 @@
 import os
 import sys
 
+try:
+    import transformers
+except ImportError:
+    raise ImportError("The 'transformers' module was not found.'")
```
Comment (nskool) [inline test/resources/mnist/model_cpu/code/mnist.py:22]: Curious why this explicit try-catch is necessary here.

File: test/integration/__init__.py [hunk 5]
```diff
@@ -85,6 +98,26 @@
...
+    resnet18_script,
+    os.path.join(resnet18_model_dir, "model.pt"),
+    resnet18_model_dir,
+    filename="resnet18.tar.gz",
+    script_path="code"
+)
+
+traced_resnet18_model_dir = os.path.join(mme_path, traced_resnet18_sub_dir)
+traced_resnet18_script = os.path.join(traced_resnet18_model_dir, code_sub_dir, "inference.py")
+traced_resnet18_tar = file_utils.make_tarfile(
```
[the remark this answers is not in the document]
Comment (sachanub) [reply test/integration/__init__.py]: Acknowledged. Updated in latest commit.

File: test/integration/local/test_multi_model_endpoint_local.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+INVOCATION_URL = BASE_URL + "models/{}/invoke"
+MODELS_URL = BASE_URL + "models"
+DELETE_MODEL_URL = BASE_URL + "models/{}"
+
+
+@pytest.fixture(scope="module", autouse=True)
+def container(image_uri, use_gpu):
+    try:
+        gpu_option = "--gpus device=0" if use_gpu else ""
+        resnet18_path = os.path.join(mme_path, 'resnet18')
```
[the remark this answers is not in the document]
Comment (sachanub) [reply test/integration/local/test_multi_model_endpoint_local.py]: Acknowledged. Updated in latest commit.

File: test/resources/mnist/model_cpu/code/mnist.py (hunk 4 above)
Comment (sachanub) [reply test/resources/mnist/model_cpu/code/mnist.py]: Acknowledged. Removed try-catch block. If `transformers` is not present, the error will show up anyways. Added comment above `import` statement to explain the reason for importing `transformers` i.e. to check that dependencies in `requirements.txt` get installed.

**Review by nskool — APPROVED** [review]

**Review by chen3933 — APPROVED** [review]

**Pull request merged**